### PR TITLE
Deployment users query

### PR DESCRIPTION
**Pull request by kushalmalani**
## Description

> Describe the purpose of this pull request.
query to fetch all users associated with a deployment

## 🎟 Issue(s)

Resolves astronomer/issues#1762

## 🧪 Functional Testing

> List the functional testing steps to confirm this feature or fix.

## 📸 Screenshots

> Add screenshots to illustrate the validity of these changes.

## 📋 Checklist

- [ ] Rebased from the main (or release if patching) branch (before testing)
- [ ] Ran `yarn lint && yarn test` before taking out of draft
- [ ] Added/updated applicable tests
- [ ] Tested against [Astro-UI](https://github.com/astronomer/astro-ui/) and [Astro-CLI](https://github.com/astronomer/astro-cli/) clients.
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

### CODE REVIEW

File: src/resolvers/query/deployment-users/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { userQuery } from "users";
+import { USER_STATUS_ACTIVE } from "constants";
+/*
+ * Get list of deployment users.
+ * @param {Object} parent The result of the parent resolver.
+ * @param {Object} args The graphql arguments.
+ * @param {Object} ctx The graphql context.
+ * @return users on this specific deployment
+ */
+export default async function deploymentUsers(parent, args, ctx) {
```
Comment (andriisoldatenko) [inline src/resolvers/query/deployment-users/index.js:10]: is it possible to return more than 1 user in `deploymentUsers`?
i was wondering to rename to `deploymentUser`

Comment (kushalmalani) [reply src/resolvers/query/deployment-users/index.js:10]: Yes, a deployment could have more than one user assigned to it right. I followed the same convention we have for workspaceUsers query

File: src/schema/query.js [hunk 2]
```diff
@@ -121,6 +121,22 @@ export default queryType({
       resolve: (root, args, context) => queries.deployments(root, args, context)
     });
 
+    t.field("deploymentUsers", {
+      type: "User",
+      list: [false],
+      nullable: true,
```
Comment (andriisoldatenko) [inline src/schema/query.js:127]: please add:
`description `

File: src/schema/query.js [hunk 3]
```diff
@@ -121,6 +121,22 @@ export default queryType({
...
+      args: {
+        deploymentUuid: arg({ type: "Uuid", required: true }),
+        user: arg({
+          type: "UserSearch",
+          nullable: true,
+          description: "A `UserSearch` input type."
```
Comment (andriisoldatenko) [inline src/schema/query.js:133]: 👍

File: src/schema/query.js [hunk 4]
```diff
@@ -121,6 +121,22 @@ export default queryType({
       resolve: (root, args, context) => queries.deployments(root, args, context)
     });
 
+    t.field("deploymentUsers", {
+      type: "User",
+      list: [false],
+      nullable: true,
+      args: {
+        deploymentUuid: arg({ type: "Uuid", required: true }),
```
Comment (andriisoldatenko) [inline src/schema/query.js:129]: please add `description`

File: src/resolvers/query/deployment-users/index.unit.test.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+          some: {
+            deployment: { id: vars.deploymentUuid }
+          }
+        }
+      },
+      include: { emails: true }
+    });
+  });
+
+  test("typical (single user) request is successful", async () => {
```
Comment (andriisoldatenko) [inline src/resolvers/query/deployment-users/index.unit.test.js:68]: can we add few negative tests.

**Comment by andriisoldatenko** [comment]
@kushalmalani i created 1 deployment, and i have 1 user (andrii@astronomer.io) - sysadmin:

```
query deploymentUsers {
  deploymentUsers(deploymentId: "ckgqokp010395opmqf5ttb599", user:{email: "andrii@astronomer.io"}) {
    username
    fullName
  }
}

{
  "data": {
    "deploymentUsers": []
  }
}
```

does it work as expected? or should we return me as user?

cc @adam2k

**Comment by kushalmalani** [comment]
> @kushalmalani i created 1 deployment, and i have 1 user ([andrii@astronomer.io](mailto:andrii@astronomer.io)) - sysadmin:
> 
> ```
> query deploymentUsers {
>   deploymentUsers(deploymentId: "ckgqokp010395opmqf5ttb599", user:{email: "andrii@astronomer.io"}) {
>     username
>     fullName
>   }
> }
> 
> {
>   "data": {
>     "deploymentUsers": []
>   }
> }
> ```
> 
> does it work as expected? or should we return me as user?
> 
> cc @adam2k

@andriisoldatenko @adam2k Yes, the above query is right and it should return you the system-admin user. That will work, once i remove the code for synthetic rolebindings and actually create a deployment role binding when a new deployment is created. That will come as part of next PR.

The current changes are new query and mutations and doesnt effect anything on merge to dev. But the changes in next PR are to existing mutations and all that have to go in a single push to dev, after tested and code reviewed.

**Comment by adam2k** [comment]
> @kushalmalani i created 1 deployment, and i have 1 user ([andrii@astronomer.io](mailto:andrii@astronomer.io)) - sysadmin:
> 
> ```
> query deploymentUsers {
>   deploymentUsers(deploymentId: "ckgqokp010395opmqf5ttb599", user:{email: "andrii@astronomer.io"}) {
>     username
>     fullName
>   }
> }
> 
> {
>   "data": {
>     "deploymentUsers": []
>   }
> }
> ```
> 
> does it work as expected? or should we return me as user?
> 
> cc @adam2k

@andriisoldatenko I think this is actually OK.  Take this use case:

A customer explicitly gives access to 1 person within their organization, and they they query to get the list of users `astro deployment user ls` we would want them to only see the user that they added and not all the SysAdmins.  I'd imagine this would confuse people especially on prod.  

If somebody was to see a bunch of our support team listed or other `@astronomer.io` addresses I'm guessing they would submit a support ticket.

**Comment by andriisoldatenko** [comment]
@adam2k i think question is why i can't see myself, not all sysadmins. So it's special case i think.
we should compare similar feature in workspace tab.

cc @schnie wdyt?

**Comment by kushalmalani** [comment]
> @adam2k i think question is why i can't see myself, not all sysadmins. So it's special case i think.
> we should compare similar feature in workspace tab.
> 
> cc @schnie wdyt?

You cannot see yourself now because there is no deployment role binding for that user. We dont create that role binding in today's system. You should see that, once the next PR change comes in.

**Comment by kushalmalani** [comment]
https://github.com/astronomer/houston-api/pull/522

**Pull request closed without merging**